packages/ampd-sdk/src/grpc/connection_manager.rs

```diff
@@ -0,0 +1,257 @@
+///
+/// This is a multi-producer single consumer pattern:
+/// - The manager maintains the connection state via a watch channel
+/// - Multiple clients can subscribe to connection state changes
+/// - Clients communicate back to the manager via mpsc channel
+/// - The manager handles connection retries and failure recovery
+#[derive(Debug)]
+pub struct ConnectionManager {
+    url: Url,
+    connection_state: watch::Sender<ConnectionState>,
+    message_receiver: mpsc::Receiver<ClientMessage>,
+    retry_policy: RetryPolicy,
+}
+
+impl ConnectionManager {
+    pub fn new(url: &str) -> Result<(Self, ConnectionHandle), Error> {
+        let url = Url::new_sensitive(url).change_context(AppError::InvalidUrl.into())?;
+
+        let (state_sender, state_receiver) = watch::channel(ConnectionState::Disconnected);
+        let (msg_sender, msg_receiver) = mpsc::channel(100);
+
+        let handle = ConnectionHandle {
+            connection_receiver: state_receiver,
+            message_sender: msg_sender,
+        };
+
+        let manager = Self {
+            url,
+            connection_state: state_sender,
+            message_receiver: msg_receiver,
+            retry_policy: DEFAULT_RETRY_POLICY,
+        };
+
+        Ok((manager, handle))
+    }
+
+    pub async fn run(mut self) -> Result<(), Error> {
+        if let Err(e) = self.connect().await {
+            warn!(err = ?e, "initial connection failed, will retry on client requests");
+        }
+
+        while let Some(message) = self.message_receiver.recv().await {
+            match message {
+                ClientMessage::ConnectionFailed(details) => {
+                    warn!("client reported connection failure: {}", details);
+                    self.handle_connection_failure().await?
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn connect(&self) -> Result<(), Error> {
+        let endpoint: transport::Endpoint = self.url.as_str().parse().into_report()?;
+        let endpoint = endpoint
+            .connect_timeout(DEFAULT_INITIAL_TIMEOUT)
+            .timeout(DEFAULT_RPC_TIMEOUT)
+            .keep_alive_timeout(KEEPALIVE_TIMEOUT)
+            .keep_alive_while_idle(KEEPALIVE_WHILE_IDLE)
+            .http2_keep_alive_interval(KEEPALIVE_TIME);
+
+        match endpoint.connect().await {
+            Ok(channel) => {
+                self.connection_state
+                    .send(ConnectionState::Connected(channel))
+                    .into_report()?;
+                Ok(())
+            }
+            Err(status) => {
+                let status_error = Err(status).into_report();
+                if self
+                    .connection_state
+                    .send(ConnectionState::Disconnected)
+                    .is_err()
+                {
+                    return status_error
+                        .attach_printable("connection failed and could not notify clients");
```

Comment:
don't use attach_printable for something that you could just as well do with change_context. In fact, prefer not to use attach_printable at all anymore in favour of #[instrument] if you need access to additional data.